src/m365/spo/commands/tenant/tenant-applicationcustomizer-add.ts

```diff
@@ -0,0 +1,189 @@
+import commands from '../../commands';
+import Command from '../../../../Command';
+import { Cli, CommandOutput } from '../../../../cli/Cli';
+import { Options as spoListItemAddCommandOptions } from '../listitem/listitem-add';
+import { Options as spoListItemListCommandOptions } from '../listitem/listitem-list';
+import * as spoTenantAppCatalogUrlGetCommand from '../tenant/tenant-appcatalogurl-get';
+import * as spoListItemAddCommand from '../listitem/listitem-add';
+import * as spoListItemListCommand from '../listitem/listitem-list';
+import { urlUtil } from '../../../../utils/urlUtil';
+
+interface Solution {
+  ContainsTenantWideExtension: boolean;
+  SkipFeatureDeployment: boolean;
+}
+
+interface CommandArgs {
+  options: Options;
+}
+
+interface Options extends GlobalOptions {
+  title: string;
+  clientSideComponentId: string;
+  clientSideComponentProperties?: string;
+  webTemplate?: string;
+}
+
+class SpoTenantApplicationCustomizerAddCommand extends SpoCommand {
+  public get name(): string {
+    return commands.TENANT_APPLICATIONCUSTOMIZER_ADD;
+  }
+
+  public get description(): string {
+    return 'Add an application customizer as a tenant wide extension.';
+  }
+
+  constructor() {
+    super();
+
+    this.#initTelemetry();
+    this.#initOptions();
+    this.#initValidators();
+  }
+
+  #initTelemetry(): void {
+    this.telemetry.push((args: CommandArgs) => {
+      Object.assign(this.telemetryProperties, {
+        clientSideComponentProperties: typeof args.options.clientSideComponentProperties !== 'undefined',
+        webTemplate: typeof args.options.webTemplate !== 'undefined'
+      });
+    });
+  }
+
+  #initOptions(): void {
+    this.options.unshift(
+      {
+        option: '-t, --title <title>'
+      },
+      {
+        option: '-i, --clientSideComponentId <clientSideComponentId>'
+      },
+      {
+        option: '--clientSideComponentProperties [clientSideComponentProperties]'
+      },
+      {
+        option: '--webTemplate [webTemplate]'
+      }
+    );
+  }
+
+  #initValidators(): void {
+    this.validators.push(
+      async (args: CommandArgs) => {
+        if (!validation.isValidGuid(args.options.clientSideComponentId)) {
+          return `${args.options.clientSideComponentId} is not a valid GUID`;
+        }
+
+        return true;
+      }
+    );
+  }
+
+  public async commandAction(logger: Logger, args: CommandArgs): Promise<void> {
+    try {
+      if (this.verbose) {
+        logger.logToStderr(`Adding application customizer with clientSideComponentId ${args.options.clientSideComponentId} and title ${args.options.title} to the tenant app catalog.`);
+      }
+      const appCatalogUrl = await this.getAppCatalogUrl();
+      if (this.verbose) {
+        logger.logToStderr(`Got tenant app catalog url: ${appCatalogUrl}`);
+      }
+
+      const solutionId = await this.getSolutionIdFromComponentManifestList(appCatalogUrl, args.options.clientSideComponentId, logger);
```

Comment:
The only thing we are currently not checking, is if the manifest that we're retrieving is in fact an ApplicationCustomizer. This information is in the JSON manifest.

Let's retrieve this together with the SolutionId so we can validate that as well.

Reply:
Great suggestion. Added this.

Reply:
Would it be slightly better if we would return the componentManifest from the function? You could rename the `getSolutionIdFromComponentManifestList` function to `getComponentManifest`, and validate the manifest in this central function. That way we have all validations in one place. The reason to do this would also be single responsibility: have the get function just to a get instead of a get + validation checks.